### PR TITLE
Added block.MetaFetcher logic for resilient sync of meta files.

### PR DESCRIPTION
**Pull request by bwplotka**
This is meant to replace least 5 inconsistent meta.json syncs places across components.

Usage of those will be done in follow up chained PRs.

In the first commit: Transactional gauge allowing to have atomic gauge across different labels.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/block/fetcher.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+	defer runutil.CloseWithLogOnErr(s.logger, r, "close bkt meta get")
+
+	metaContent, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read meta file: %v", metaFile)
+	}
+
+	m := &metadata.Meta{}
+	if err := json.Unmarshal(metaContent, m); err != nil {
+		return nil, errors.Wrapf(ErrorSyncMetaCorrupted, "meta.json %v unmarshal: %v ", metaFile, err)
```
Comment (GiedriusS) [inline pkg/block/fetcher.go]: Small nit: that space at the end seems unnecessary. The same happened on some other lines as well.

File: pkg/block/fetcher.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+
+	corruptedMeta = "corrupted-meta-json"
+	noMeta        = "no-meta-json"
+	loadedMeta    = "loaded"
+	failedMeta    = "failed"
+
+	// Filter's label values.
+	labelExcludedMeta = "label-excluded"
+	timeExcludedMeta  = "time-excluded"
+	TooFreshMeta      = "too-fresh"
```
Comment (GiedriusS) [inline pkg/block/fetcher.go]: Does this need to be exposed?

File: pkg/block/fetcher.go [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+	for id, m := range metas {
+		if m.MaxTime >= f.minTime.PrometheusTimestamp() && m.MinTime <= f.maxTime.PrometheusTimestamp() {
+			continue
+		}
+		synced.WithLabelValues(timeExcludedMeta).Inc()
+		delete(metas, id)
+	}
+}
+
+var _ MetaFetcherFilter = (&LabelShardedMetaFilter{}).Filter
```
Comment (GiedriusS) [inline pkg/block/fetcher.go]: Also lets not forget to remove this and the other one. I presume some linter was complaining 😄

Comment (bwplotka) [reply pkg/block/fetcher.go]: Why removing this? (: This is ensuring Filter meets interface

File: pkg/block/fetcher.go (hunk 2 above)
Comment (bwplotka) [reply pkg/block/fetcher.go]: Yes - used in https://github.com/thanos-io/thanos/pull/1937 (:

File: pkg/block/fetcher.go (hunk 3 above)
Comment (GiedriusS) [reply pkg/block/fetcher.go]: Ah, right, that's ok, I guess :+1:

**Pull request merged**